### PR TITLE
week 4 assessment complete

### PR DESCRIPTION
**Pull request by crgjr66**

### CODE REVIEW

File: code-challenges.js [hunk 1]
```diff
@@ -6,9 +6,22 @@ var collections = ["purple", "blue", "green", "yellow", "pink"]
...
+      var temp = array[i];
+      array[i] = array[j];
+      array[j] = temp;
+  }
+  return array
+}
+
+console.log(shrinkAndShuffle(collections))
+console.log(shrinkAndShuffle(collections))
+console.log(shrinkAndShuffle(collections))
```
Comment (sjproctor) [inline code-challenges.js:25]: Interesting approach!

File: code-challenges.js [hunk 2]
```diff
@@ -17,9 +30,20 @@ var cubeAndSum1 = [2, 3, 4]
...
+  // a number defaulted to 0
+  return array.map(value => value**3).reduce((a, b) => a + b, 0)
+  // for (let i = 0; i < array.length; i++) {
+  //   sum += array[i]
+  // }
+  // return sum
+}
+
+console.log(sumCubed(cubeAndSum1))
+console.log(sumCubed(cubeAndSum2))
```
Comment (sjproctor) [inline code-challenges.js:47]: Love the reduce approach. It is handy method.

File: code-challenges.js [hunk 3]
```diff
@@ -28,9 +52,15 @@ var nums1 = [3, 56, 90, -8, 0, 23, 6]
...
 var nums2 = [109, 5, 9, -59, 8, 24]
 // Expected output: [-59, 109]
+const findMinMax = (array) => {
+  let new_array = []
+  new_array.push(Math.min(...array))
+  new_array.push(Math.max(...array))
+  return new_array
+}
 
-
```
Comment (sjproctor) [inline code-challenges.js:32]: Love it! Super concise.

File: code-challenges.js [hunk 4]
```diff
@@ -39,13 +69,47 @@ var testString1 = "albatross"
...
+  // let new_string = string.split("");
+  // console.log(new_string);
+  // for (let i = 1; i < new_string.length; i += 2) {
+  //   new_string[i] = new_string[i].toUpperCase();
+  // }
+  // return new_string.join("");
+}
+
+console.log(altCapLetter(testString1))
+console.log(altCapLetter(testString2))
```
Comment (sjproctor) [inline code-challenges.js:91]: I am curious, is the for loop a first attempt or just a separate approach? Is there a reason for leaving the commented-out code?

File: code-challenges.js [hunk 5]
```diff
@@ -39,13 +69,47 @@ var testString1 = "albatross"
...
+    for (let j = 0; j < new_array.length; j++) {
+      if (!new_array.includes(value)) {
+        new_array.push(value);
+      }
+    }
+  })
+  return new_array;
+}
+
+console.log(noDups(arr1, arr2))
```
Comment (sjproctor) [inline code-challenges.js:115]: Interesting that you used a for loop inside a map. I think you are over engineering that part a bit.

File: code_challenges.rb [hunk 6]
```diff
@@ -6,42 +6,133 @@
...
+  new_array = []
+  array.each do |value|
+    if value.class == Integer && value % 2 == 1
+      new_array << value
+    end
+  end
+  new_array.sort
+end
 
-
```
Comment (sjproctor) [inline code_challenges.rb:10]: Love it!

File: code_challenges.rb [hunk 7]
```diff
@@ -6,42 +6,133 @@
...
+  end
 
+  def bike_info
+    "The #{@model} bike has #{@wheels} wheels and a #{@frame_size} frame."
+  end
+end
 
-
+my_bike = Bike.new
+p my_bike.bike_info
```
Comment (sjproctor) [inline code_challenges.rb:38]: All your logic is right on. I would consider passing the model and frame in on initialization so you are not limited to one type of bike object.

File: code_challenges.rb [hunk 8]
```diff
@@ -6,42 +6,133 @@
...
+      @speed -= speed
+    else
+      @speed = 0
+    end
+  end
+end
+
+my_everything_bike = Bike4.new
+p my_everything_bike.pedal_faster 10
+p my_everything_bike.brake 15
```
Comment (sjproctor) [inline code_challenges.rb:136]: Looks awesome!

File: interview-questions.md [hunk 9]
```diff
@@ -7,56 +7,58 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
+  Researched answer: Whenever a code block is executed Ruby returns the value of the last expression. An example is the getter method. We call a method and always expect some value in return. The keyword 'return' is not necessary.
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
-
-  Researched answer:
+  Your answer: A programming method that is based on objects, classes, and inheritance, rather than functions and logic (i.e. the way I write code).
```
Comment (sjproctor) [inline interview-questions.md:33]: Yup, me too! haha

**Review by sjproctor — APPROVED** [review]
Nice work! Great short answers and thorough coding. Keep it up!

**Comment by crgjr66** [comment]
The reason for leaving in the comments is because I still think in terms of lower level mathematical functionality, not javascript API's and higher order functions.  In order for me to solve the problem, I still code the logic and get it working first, then go back and research the higher order function in order to streamline the solution.  I am hoping this will change over time as I become more familiar with all the possible javascript functions available to use.  I was also confused in Monday's review of the assessment, when we were discouraged from using Math.min/max to find those values in an array; citing that the lengthier, logical mathematical solution is preferred during an interview.  Conceptually and developmentally my coding is still in a hybrid phase.

**Pull request merged**